android/app/src/main/java/com/sample/react/TestAppReactNativeHost.kt

```diff
@@ -2,45 +2,69 @@ package com.sample.react
 
 import android.app.Activity
 import android.app.Application
-import com.facebook.react.*
+import com.facebook.react.PackageList
+import com.facebook.react.ReactInstanceManager
+import com.facebook.react.ReactNativeHost
+import com.facebook.react.ReactPackage
 import com.facebook.react.bridge.ReactMarker
 import com.facebook.react.bridge.ReactMarkerConstants
 import com.facebook.react.common.LifecycleState
 import com.facebook.soloader.SoLoader
 import javax.inject.Inject
 import javax.inject.Singleton
 
+sealed class BundleSource {
+    enum class Action {
+        RESTART, RELOAD
+    }
+
+    abstract fun moveTo(to: BundleSource): Action
+
+    object Disk : BundleSource() {
+        override fun moveTo(to: BundleSource) = Action.RESTART
+    }
+
+    object Server : BundleSource() {
+        override fun moveTo(to: BundleSource): Action {
+            return when (to) {
+                Disk -> Action.RESTART
+                Server -> Action.RELOAD
+            }
+        }
+    }
+}
+
 @Singleton
 class TestAppReactNativeHost @Inject constructor(
-        application: Application,
-        private val reactBundleNameProvider: ReactBundleNameProvider
+    application: Application,
+    private val reactBundleNameProvider: ReactBundleNameProvider
 ) : ReactNativeHost(application) {
+    var source: BundleSource = reactBundleNameProvider.bundleName
+        ?.let { BundleSource.Disk } ?: BundleSource.Server
 
-    private var currentActivity: Activity? = null
-    private var useEmbeddedBundle: Boolean = true
-
-    fun reload(activity: Activity, useEmbeddedBundle: Boolean) {
+    fun reload(activity: Activity, newSource: BundleSource) {
         assert(hasInstance()) {
-            "startInBackground() must be called the first time ReactInstanceManager is created"
-        }
-
-        if (!useEmbeddedBundle && useEmbeddedBundle == this.useEmbeddedBundle) {
-            reactInstanceManager.devSupportManager.handleReloadJS()
-            return
+            "init() must be called the first time ReactInstanceManager is created"
         }
 
-        clear()
+        val action = source.moveTo(newSource)
+        source = newSource
 
-        this.currentActivity = activity
-        this.useEmbeddedBundle = useEmbeddedBundle
+        when (action) {
+            BundleSource.Action.RELOAD -> reactInstanceManager.devSupportManager.handleReloadJS()
```

Comment:
Nit: Can we add braces to make the block consistent?
```suggestion
            BundleSource.Action.RELOAD -> {
                reactInstanceManager.devSupportManager.handleReloadJS()
            }
```

Reply:
Done!